src/rbtree.rs

```diff
@@ -95,422 +154,646 @@ impl fmt::Debug for Link {
-        };
-        let mut x = if !y.left().is_null() {
-            y.left()
-        } else {
-            y.right()
-        };
-        let mut w = NodePtr::null();
-        if !x.is_null() {
-            x.set_parent(y.parent());
-        }
-        if y.parent().is_null() {
-            *root = x;
-        } else if y.is_left_child() {
-            y.parent().set_left(x);
-            w = y.parent().right();
-        } else {
-            y.parent().set_right(x);
-            w = y.parent().left();
-        }
-        let removed_black = y.color() == Color::Black;
-        if y != self {
-            y.set_parent(self.parent());
-            if self.parent().is_null() {
-                *root = y;
-            } else if self.is_left_child() {
-                y.parent().set_left(y);
+            link_ops.set_right(y_parent, x);
+            w = link_ops.left(y_parent);
+        }
+    } else {
+        *root = x;
+    }
+    let removed_black = link_ops.color(y) == Color::Black;
+    if y != ptr {
+        if let Some(parent) = link_ops.parent(ptr) {
+            link_ops.set_parent(y, Some(parent));
+            if is_left_child(link_ops, ptr, parent) {
+                link_ops.set_left(parent, Some(y));
             } else {
-                y.parent().set_right(y);
-            }
-            y.set_left(self.left());
-            y.left().set_parent(y);
-            y.set_right(self.right());
-            if !y.right().is_null() {
-                y.right().set_parent(y);
+                link_ops.set_right(parent, Some(y));
             }
-            y.set_color(self.color());
+        } else {
+            link_ops.set_parent(y, None);
+            *root = Some(y);
         }
-        if removed_black && !root.is_null() {
-            if !x.is_null() {
-                x.set_color(Color::Black);
-            } else {
-                loop {
-                    if !w.is_left_child() {
-                        if w.color() == Color::Red {
-                            w.set_color(Color::Black);
-                            w.parent().set_color(Color::Red);
-                            w.parent().rotate_left(root);
-                            w = w.left().right();
-                        }
-                        if (w.left().is_null() || w.left().color() == Color::Black)
-                            && (w.right().is_null() || w.right().color() == Color::Black)
+        link_ops.set_left(y, link_ops.left(ptr));
+        link_ops.set_parent(link_ops.left(y).unwrap_unchecked(), Some(y));
+        link_ops.set_right(y, link_ops.right(ptr));
+        if !link_ops.right(y).is_none() {
+            link_ops.set_parent(link_ops.right(y).unwrap_unchecked(), Some(y));
+        }
+        link_ops.set_color(y, link_ops.color(ptr));
+    }
+    if removed_black && !root.is_none() {
+        if let Some(x) = x {
+            link_ops.set_color(x, Color::Black);
+        } else {
+            loop {
+                if !is_left_child(
```

Comment:
`w.unwrapped_unchecked()` is repeated a lot in this code. Generally if you find yourself repeating this kind of code a lot you can just unwrap it once:
```rust
let w = w.unwrapped_unchecked();
```